cub/cub/device/dispatch/kernels/histogram.cuh

```diff
@@ -17,6 +17,308 @@
+    }
+
+    template <typename T, ::cuda::std::enable_if_t<!is_integral_excl_int128<T>::value, int> = 0>
+    _CCCL_HOST_DEVICE _CCCL_FORCEINLINE int ComputeBin(T sample, T min_level, ScaleT scale)
+    {
+      return this->ComputeBin(sample, min_level, scale, ::cuda::std::is_floating_point<T>{});
+    }
+
+#if _CCCL_HAS_NVFP16()
+    _CCCL_HOST_DEVICE _CCCL_FORCEINLINE int ComputeBin(__half sample, __half min_level, ScaleT scale)
+    {
+      NV_IF_TARGET(
+        NV_PROVIDES_SM_53,
+        (return static_cast<int>(__hmul(__hsub(sample, min_level), scale.reciprocal));),
+        (return static_cast<int>((__half2float(sample) - __half2float(min_level)) * __half2float(scale.reciprocal));));
+    }
+#endif // _CCCL_HAS_NVFP16()
+
+    _CCCL_HOST_DEVICE _CCCL_FORCEINLINE bool
+    MayOverflow(CommonT /* num_bins */, ::cuda::std::false_type /* is_integral */)
+    {
+      return false;
+    }
+
+    /**
+     * @brief Returns true if the bin computation for a given combination of range `(max_level -
+     * min_level)` and number of bins may overflow.
+     */
+    _CCCL_HOST_DEVICE _CCCL_FORCEINLINE bool MayOverflow(CommonT num_bins, ::cuda::std::true_type /* is_integral */)
+    {
+      return static_cast<IntArithmeticT>(m_max - m_min)
+           > (::cuda::std::numeric_limits<IntArithmeticT>::max() / static_cast<IntArithmeticT>(num_bins));
+    }
+
+  public:
+    /**
+     * @brief Initializes the ScaleTransform for the given parameters
+     * @return cudaErrorInvalidValue if the ScaleTransform for the given values may overflow,
+     * cudaSuccess otherwise
+     */
+    _CCCL_HOST_DEVICE _CCCL_FORCEINLINE cudaError_t Init(int num_levels, LevelT max_level, LevelT min_level)
+    {
+      m_max = static_cast<CommonT>(max_level);
+      m_min = static_cast<CommonT>(min_level);
+
+      // Check whether accurate bin computation for an integral sample type may overflow
+      if (MayOverflow(static_cast<CommonT>(num_levels - 1), ::cuda::std::is_integral<CommonT>{}))
+      {
+        return cudaErrorInvalidValue;
+      }
+
+      m_scale = this->ComputeScale(num_levels, m_max, m_min);
+      return cudaSuccess;
+    }
+
+    // Method for converting samples to bin-ids
+    template <CacheLoadModifier LOAD_MODIFIER>
+    _CCCL_HOST_DEVICE _CCCL_FORCEINLINE void BinSelect(SampleT sample, int& bin, bool valid)
+    {
+      const CommonT common_sample = static_cast<CommonT>(sample);
+
+      if (valid && this->SampleIsValid(common_sample, m_max, m_min))
+      {
+        bin = this->ComputeBin(common_sample, m_min, m_scale);
+      }
+    }
+  };
+
+  // Pass-through bin transform operator
+  struct PassThruTransform
+  {
+// GCC 14 rightfully warns that when a value-initialized array of this struct is copied using memcpy, uninitialized
+// bytes may be accessed. To avoid this, we add a dummy member, so value initialization actually initializes the memory.
+#if _CCCL_COMPILER(GCC, >=, 13)
+    char dummy;
+#endif
```

Comment:
I would prefer if we did not change class layout based on compiler.

Reply:
This was already in the code, I just moved it to this file

Reply:
I introduced this in #2157.